apps/prairielearn/src/lib/bot-grading.ts

```diff
@@ -0,0 +1,182 @@
+    let output_count = 0;
+    let output: string | null = null;
+
+    // get each instance question
+    for (const instance_question of result) {
+      const { variant, submission } = await queryRow(
+        sql.select_last_variant_and_submission,
+        { instance_question_id: instance_question.id },
+        SubmissionVariantSchema,
+      );
+
+      // build urls for the question server
+      const urls = buildQuestionUrls(urlPrefix, variant, question, instance_question);
+
+      // get question html
+      const questionModule = questionServers.getModule(question.type);
+      const { courseIssues, data } = await questionModule.render(
+        { question: true, submissions: false, answer: false },
+        variant,
+        question,
+        null,
+        [],
+        question_course,
+        urls,
+      );
+      if (courseIssues.length) {
+        job.info(courseIssues.toString());
+        job.error('Error occurred');
+        job.fail('Errors occurred while bot grading, see output for details');
+      }
+      const question_prompt_raw = data.questionHtml;
+      const $ = cheerio.load(question_prompt_raw, null, false);
+      $('script').remove();
+      const question_prompt = $.html();
+
+      const render_submission_results = await questionModule.render(
+        { question: false, submissions: true, answer: false },
+        variant,
+        question,
+        submission,
+        [submission],
+        question_course,
+        urls,
+      );
+      const student_answer = render_submission_results.data.submissionHtmls[0];
+
+      // Call OpenAI API
+      const completion = await openai.chat.completions.create({
+        messages: [
+          {
+            role: 'system',
+            content: `You are an instructor for a course, and you are grading assignments. You should always return the grade using a json object of 2 parameters: grade and feedback. The grade should be an integer between 0 and 100. 0 being the lowest and 100 being the highest, and the feedback should be why you give this grade, or how to improve the answer. You can say correct or leave blank when the grade is close to 100. `,
+          },
+          {
+            role: 'user',
+            content: `Question: \n${question_prompt} \nAnswer: \n${student_answer} \nHow would you grade this? Please return the json object.`,
+          },
+        ],
+        model: 'gpt-3.5-turbo',
+      });
+
+      let msg = '';
+      try {
+        if (completion.choices[0].message.content === null) {
+          error_count++;
+          continue;
+        }
+        const gpt_answer = JSON.parse(completion.choices[0].message.content);
+        await manualGrading.updateInstanceQuestionScore(
+          assessment_question.assessment_id,
+          instance_question.id,
+          submission.id,
+          null, // modified_at
+          {
+            score_perc: gpt_answer.grade,
+            feedback: { manual: gpt_answer.feedback },
```

Comment:
Do you want to consider validating the `gpt_answer` object? You could do this pretty easily with a Zod schema.